lib/wasi/src/runtime/resolver/types.rs

```diff
@@ -0,0 +1,197 @@
+use std::{collections::BTreeMap, fmt::Display, ops::Deref, path::PathBuf, str::FromStr};
+
+use crate::{
+    bin_factory::BinaryPackage,
+    http::HttpClient,
+    runtime::resolver::{cache::CacheConfig, InMemoryCache},
+};
+
+#[async_trait::async_trait]
+pub trait PackageResolver: std::fmt::Debug + Send + Sync {
+    /// Resolve a package, loading all dependencies.
+    async fn resolve_package(
+        &self,
+        pkg: &WebcIdentifier,
+        client: &(dyn HttpClient + Send + Sync),
```

Comment:
A resolver might not need an http client at all. Or might need to use a specific http client.

This shouldn't be part of the API, but passed in during specific resolver construction.

Reply:
Note that `WasiRuntime::http_client()` should also be changed to return an `Arc<>`,  or to have a `build_http_client()` instead that constructs a new, independent client.

Reply:
Yeah, I agree.

Depending on how we want to tackle this, we could a) pass in some sort of factory that'll give you a HTTP client (or TCP connection or whatever) that we got from `WasiRuntime::http_client()`, or we could b) require `PackageResolver` implementations to bring their own HTTP client.

I'm partial to the first approach because it makes the code more testable and lets the runtime manage all HTTP client instances.

Reply:
I'm going to leave this for the next PR where we start to flesh out the `PackageResolver` trait and clean up the code that uses it.